# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/wr/WebRenderCommandBuilder.cpp
# Commit: a2da895a58ce
# Full Hash: a2da895a58ce094d55e6a2e040dd648962b39868
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2022-02-22 09:37:09
# Regressor Bug: 1714584
# File Overlap Count: 2
# Description:
#   Bug 1714584 - Part 1: Decouple nsDisplayList internal list from nsDisplayItems r=mstange
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138152
# ==============================================================================

diff -r c3b3315d2344 -r a2da895a58ce gfx/layers/wr/WebRenderCommandBuilder.cpp
--- a/gfx/layers/wr/WebRenderCommandBuilder.cpp	Mon Feb 21 21:57:03 2022 +0000
+++ b/gfx/layers/wr/WebRenderCommandBuilder.cpp	Tue Feb 22 00:44:25 2022 +0000
@@ -558,7 +558,8 @@
                 nsDisplayListBuilder* aDisplayListBuilder,
                 wr::DisplayListBuilder& aBuilder,
                 wr::IpcResourceUpdateQueue& aResources, Grouper* aGrouper,
-                nsDisplayItem* aStartItem, nsDisplayItem* aEndItem) {
+                nsDisplayList::iterator aStartItem,
+                nsDisplayList::iterator aEndItem) {
     GP("\n\n");
     GP("Begin EndGroup\n");
 
@@ -654,6 +655,7 @@
 
     RenderRootStateManager* rootManager =
         aWrManager->GetRenderRootStateManager();
+
     bool empty = aStartItem == aEndItem;
     if (empty) {
       ClearImageKey(rootManager, true);
@@ -663,8 +665,8 @@
     // Reset mHitInfo, it will get updated inside PaintItemRange
     mHitInfo = CompositorHitTestInvisibleToHit;
 
-    PaintItemRange(aGrouper, nsDisplayList::Range(aStartItem, aEndItem),
-                   context, recorder, rootManager, aResources);
+    PaintItemRange(aGrouper, aStartItem, aEndItem, context, recorder,
+                   rootManager, aResources);
 
     // XXX: set this correctly perhaps using
     // aItem->GetOpaqueRegion(aDisplayListBuilder, &snapped).
@@ -769,14 +771,14 @@
                          SideBits::eNone);
   }
 
-  void PaintItemRange(Grouper* aGrouper, nsDisplayList::Iterator aIter,
-                      gfxContext* aContext,
+  void PaintItemRange(Grouper* aGrouper, nsDisplayList::iterator aStartItem,
+                      nsDisplayList::iterator aEndItem, gfxContext* aContext,
                       WebRenderDrawEventRecorder* aRecorder,
                       RenderRootStateManager* aRootManager,
                       wr::IpcResourceUpdateQueue& aResources) {
     LayerIntSize size = mVisibleRect.Size();
-    while (aIter.HasNext()) {
-      nsDisplayItem* item = aIter.GetNext();
+    for (auto it = aStartItem; it != aEndItem; ++it) {
+      nsDisplayItem* item = *it;
       MOZ_ASSERT(item);
 
       BlobItemData* data = GetBlobItemData(item);
@@ -949,7 +951,7 @@
         aContext->GetDrawTarget()->FlushItem(aItemBounds);
       } else {
         aContext->Multiply(ThebesMatrix(trans2d));
-        aGroup->PaintItemRange(this, nsDisplayList::Iterator(aChildren),
+        aGroup->PaintItemRange(this, aChildren->begin(), aChildren->end(),
                                aContext, aRecorder, aRootManager, aResources);
       }
 
@@ -972,8 +974,8 @@
       GP("beginGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
          aItem->GetPerFrameKey());
       aContext->GetDrawTarget()->FlushItem(aItemBounds);
-      aGroup->PaintItemRange(this, nsDisplayList::Iterator(aChildren), aContext,
-                             aRecorder, aRootManager, aResources);
+      aGroup->PaintItemRange(this, aChildren->begin(), aChildren->end(),
+                             aContext, aRecorder, aRootManager, aResources);
       aContext->GetDrawTarget()->PopLayer();
       GP("endGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
          aItem->GetPerFrameKey());
@@ -989,8 +991,8 @@
       GP("beginGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
          aItem->GetPerFrameKey());
       aContext->GetDrawTarget()->FlushItem(aItemBounds);
-      aGroup->PaintItemRange(this, nsDisplayList::Iterator(aChildren), aContext,
-                             aRecorder, aRootManager, aResources);
+      aGroup->PaintItemRange(this, aChildren->begin(), aChildren->end(),
+                             aContext, aRecorder, aRootManager, aResources);
       aContext->GetDrawTarget()->PopLayer();
       GP("endGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
          aItem->GetPerFrameKey());
@@ -1003,8 +1005,8 @@
       GP("beginGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
          aItem->GetPerFrameKey());
       aContext->GetDrawTarget()->FlushItem(aItemBounds);
-      aGroup->PaintItemRange(this, nsDisplayList::Iterator(aChildren), aContext,
-                             aRecorder, aRootManager, aResources);
+      aGroup->PaintItemRange(this, aChildren->begin(), aChildren->end(),
+                             aContext, aRecorder, aRootManager, aResources);
       aContext->GetDrawTarget()->PopLayer();
       GP("endGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
          aItem->GetPerFrameKey());
@@ -1023,7 +1025,7 @@
               GP("beginGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
                  aItem->GetPerFrameKey());
               aContext->GetDrawTarget()->FlushItem(aItemBounds);
-              aGroup->PaintItemRange(this, nsDisplayList::Iterator(aChildren),
+              aGroup->PaintItemRange(this, aChildren->begin(), aChildren->end(),
                                      aContext, aRecorder, aRootManager,
                                      aResources);
               GP("endGroup %s %p-%d\n", aItem->Name(), aItem->Frame(),
@@ -1061,8 +1063,8 @@
     }
 
     default:
-      aGroup->PaintItemRange(this, nsDisplayList::Iterator(aChildren), aContext,
-                             aRecorder, aRootManager, aResources);
+      aGroup->PaintItemRange(this, aChildren->begin(), aChildren->end(),
+                             aContext, aRecorder, aRootManager, aResources);
       break;
   }
 }
@@ -1186,15 +1188,18 @@
   RenderRootStateManager* manager =
       aCommandBuilder->mManager->GetRenderRootStateManager();
 
-  nsDisplayItem* startOfCurrentGroup = nullptr;
+  nsDisplayList::iterator startOfCurrentGroup = aList->end();
   DIGroup* currentGroup = aGroup;
 
   // We need to track whether we have active siblings for mixed blend mode.
   bool encounteredActiveItem = false;
 
-  for (nsDisplayItem* item : *aList) {
-    if (!startOfCurrentGroup) {
-      startOfCurrentGroup = item;
+  for (auto it = aList->begin(); it != aList->end(); ++it) {
+    nsDisplayItem* item = *it;
+    MOZ_ASSERT(item);
+
+    if (startOfCurrentGroup == aList->end()) {
+      startOfCurrentGroup = it;
     }
 
     if (IsItemProbablyActive(item, aBuilder, aResources, aSc, manager,
@@ -1250,7 +1255,7 @@
 
       currentGroup->EndGroup(aCommandBuilder->mManager, aDisplayListBuilder,
                              aBuilder, aResources, this, startOfCurrentGroup,
-                             item);
+                             it);
 
       {
         auto spaceAndClipChain =
@@ -1270,7 +1275,7 @@
         sIndent--;
       }
 
-      startOfCurrentGroup = nullptr;
+      startOfCurrentGroup = aList->end();
       currentGroup = &groupData->mFollowingGroup;
     } else {  // inactive item
       ConstructItemInsideInactive(aCommandBuilder, aBuilder, aResources,
@@ -1280,7 +1285,7 @@
 
   currentGroup->EndGroup(aCommandBuilder->mManager, aDisplayListBuilder,
                          aBuilder, aResources, this, startOfCurrentGroup,
-                         nullptr);
+                         aList->end());
 }
 
 // This does a pass over the display lists and will join the display items